File: dynamic/SellStock121.cc
/**
 * 121. Best Time to Buy and Sell Stock
 * Say you have an array for which the ith element is the price of a given stock on day i.
 * If you were only permitted to complete at most one transaction 
 * (i.e., buy one and sell one share of the stock), design an algorithm to find the maximum profit.
 * Note that you cannot sell a stock before you buy one.
 * 
 * Example:
 * Input: [7,6,4,3,1]
 * Output: 0
 * Explanation: In this case, no transaction is done, i.e. max profit = 0.
 * */
#include<vector>
#include<algorithm>
using namespace std;
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        // int n = prices.size();
        // int dp[n][2];
        // for (int i = 0; i < n; i++){
        //     if (i - 1 ==-1)
        //     {
        //        dp[i][0] = 0;
        //        //解释
        //        //dp[i][0] = max(dp[-1][0],dp[-1][1] + prices[i]);
        //        //           = max(0, 负无穷)
        //        dp[i][1] = -prices[i];
        //        //解释
        //        //dp[i][1] = max(dp[-1][1],dp[-1][0]-prices[i]);
        //        //         = max(负无穷,0- prices[i])
        //        continue;
        //     }
        //     dp[i][0] = max(dp[i-1][0], dp[i-1][1] + prices[i]);
        //     dp[i][1] = max(dp[i-1][1], -prices[i]);
        // }
        //return dp[n-1][0];

        //方式2优化 空间复杂度O(1)
        int dp_i_0 = 0,dp_i_1 = INT16_MIN;
        for(int price :prices){
            dp_i_0 = max(dp_i_0, dp_i_1 +price);
            dp_i_1 = max(dp_i_1, -price);
        }
        return dp_i_0;
    }
};

File: dynamic/sellstock121/sellstock121.go
/*
121. 买卖股票的最佳时机
给定一个数组 prices ，它的第 i 个元素 prices[i] 表示一支给定股票第 i 天的价格。

你只能选择 某一天 买入这只股票，并选择在 未来的某一个不同的日子 卖出该股票。设计一个算法来计算你所能获取的最大利润。

返回你可以从这笔交易中获取的最大利润。如果你不能获取任何利润，返回 0 。
*/
package sellstock121

func maxProfit(prices []int) int {
	n := len(prices)
	if n == 0 {
		return 0
	}
	dp := make([][]int, n)
	for i := range dp { //内层定义是否交易
		dp[i] = make([]int, 2)
	}
	dp[0][0] = 0
	dp[0][1] = -prices[0]
	for i := 1; i < n; i++ {
		dp[i][0] = max(dp[i-1][0], dp[i-1][1]+prices[i])
		//dp[i][1] = max(dp[i-1][1], dp[i-1][0]-prices[i]) //因为只做一次交易 那么dp[i-1][0] 不会存在
		dp[i][1] = max(dp[i-1][1], -prices[i])
	}
	return dp[n-1][0]
}
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
